lib/Analysis/FIRRTLInstanceInfo.cpp

```diff
@@ -0,0 +1,128 @@
+//===- FIRRTLInstanceInfo.cpp - Instance info analysis ----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the InstanceInfo analysis.  This is an analysis that
+// depends on the InstanceGraph analysis, but provides additional information
+// about FIRRTL operations.  This is useful if you find yourself needing to
+// selectively iterate over parts of the design.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Analysis/FIRRTLInstanceInfo.h"
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/Support/Debug.h"
+
+#ifndef NDEBUG
+#include "llvm/ADT/DepthFirstIterator.h"
+#endif
+
+#define DEBUG_TYPE "firrtl-analysis-instanceinfo"
+
+using namespace circt;
+using namespace firrtl;
+
+void InstanceInfo::LatticeValue::mergeIn(LatticeValue that) {
+  if (kind > that.kind)
+    return;
+
+  if (kind < that.kind) {
+    kind = that.kind;
+    constant = that.constant;
+    return;
+  }
+
+  if (kind == Constant && constant != that.constant)
+    kind = Mixed;
+}
+
+InstanceInfo::InstanceInfo(Operation *op, mlir::AnalysisManager &am) {
+  auto &iGraph = am.getAnalysis<InstanceGraph>();
+
+  // Visit modules in reverse post-order (visit parents before children) because
+  // information flows in this direction---the attributes of modules are
+  // determinend by their instantiations.
+  DenseSet<InstanceGraphNode *> visited;
+  for (auto *root : iGraph) {
+    for (auto *modIt : llvm::inverse_post_order_ext(root, visited)) {
```

Comment:
Presumably reverse of post order VS post-order on the inverse graph.

Good question! RPO is classic/ideal here.

Without cycles I'm not sure it matters (but would want to stare at it more...), and with them we can't guarantee convergence in a single walk (FWIW).

Using RPO, might be simplest conceptually as it's what we all expect here anyway and is more familiar for this sort of thing.

I was reaching for RPO in my suggestion, tbh.

FWIW we use inv post order elsewhere so hopefully thats correct/intentional for the same reason mentioned above, but might be good to check or ping those authors :).

Reply:
Empirically, I think it's the same in how it's being used here. Possibly, `ReversePostOrder` is slightly cleaner as it avoids having to do the pattern of nesting an `inverse_post_order_ext` under iteration over the nodes in the graph. I.e., the below can be replaced with what follows it:

``` c++
DenseSet<InstanceGraphNode *> visited;
for (auto *root : iGraph) {
  for (auto *modIt : llvm::inverse_post_order_ext(root, visited)) {
    /* ... */
  }
}
```

``` c++
ReversePostOrder<InstanceGraph *> rpo(iGraph);
for (auto *modIt : rpo) {
  /* ... */
}
```

I switched over to this in 8f1cedfec.

Reply:
Chiming in a little late to this thread,  but yes RPO and IPO should both visit the "parent" modules before the "children" modules, so it doesn't really matter which you use, but we usually use IPO for two reasons:
1. RPO does a post-order walk, storing the elements to a SmallVector, and then walks it in reverse.  IPO might be slightly faster because it doesn't need to build a vector of all modules.
2. RPO will only visit modules which are reachable from the top level module (i.e. our instance graph trait implementation does not work well with multiple public modules), while the double loop used for IPO side steps this issue and visits all modules.

Reply:
Okay, thanks youngar , this clarifies my confusion of RPO vs IPO, sticking with IPO makes sense.

Reply:
SGTM. Changed back in 3fbec15b7.